### PR TITLE
TESTENG-221: Python Framework V0

### PR DESCRIPTION
**Pull request by IsiahG**
Version 0: Architecture/Design Review

This is in response to @DeclanCodes comments below. It did not have a reply field.

There is no need to worry. This is only a POC. 
1. Logging with a consistent format is a very basic feature of any framework. I would not describe it as "logging all over the place". It has a timestamp, the module it came from, etc. It also logs to stdout and a log file and rotates the log file when it gets to a certain size. This is all customizable. 
1. I am definitely not concerned about wrappers at this point. Yes, the final framework should have wrappers that handle waiting for elements, etc.
1. All well written tests will look the same in the end. This test was created to show the format of a test with well defined setup and cleanup methods. More importantly, the test is completely data-driven and configurable by an analyst with a simple editor. They can easily set the company url, login information and any part of the test they choose without touching code.

Finally, this is not a competition between the c# framework and python. This is a shift in direction of the technology stack of Test Engineering at Namely for website-based e2e test automation.

### CODE REVIEW

File: lib/ben_enroll_base_page.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python
+
+from lib.base_page import BasePage
+import lib.Driver as Driver
+from selenium.webdriver.common.by import By
+
+class BenEnrollBasePage(BasePage):
```
Comment (DeclanCodes) [inline lib/ben_enroll_base_page.py]: This is the second layer in an inheritance chain -- the enroll pages themselves will be the 3rd. I don't disagree with using inheritance here, but I thought we were going to favor composition?

File: pom_generate.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python
```
Comment (DeclanCodes) [inline pom_generate.py]: Using code generation for these simple page objects is great!

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+FROM python:3
+
+WORKDIR /usr/src/app
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply Dockerfile]: Nope, changed...

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+COPY ./ ./ 
+COPY ./requirements.txt ./
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply Dockerfile]: Done.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FROM python:3
+
+WORKDIR /usr/src/app
+COPY ./ ./ 
+COPY ./requirements.txt ./ 
+RUN pip install --no-cache-dir -r requirements.txt
+CMD [ "./runner.py", "--modules", "tests/test_benenroll_happypath.py" ]
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply Dockerfile]: :)... I am clearly a rookie Docker lol... thanks...

File: data/test_basic_login.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+    "company": "https://sq-lab10.s.namely.com/",
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply data/test_basic_login.json]: I think it is explicitly clear where the test is being executed. This is also customizable. If you want your test to use that format, so be it.

File: lib/Driver.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+def Initialize(webdriver_config="browserstack-ike"):
+    global Instance
+    webdriver_configs = list()
+
+    Log = Logger.Instance
+    Log.info("Starting WebDriver with webdriver_config: {}".format(
+        webdriver_config))
+
+    # Get and set WebDriver config
+    webdriver_configs = get_webdriver_configs()
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply lib/Driver.py]: get_webdriver_configs returns a list of web_driver_config(s).
You only print the one that matches the config the tester specifies in their test data file.

File: pages/home_page.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+class HomePage(BasePage):
+    def get_change_benefits(self):
+        return Driver.Instance.find_element(
+            By.XPATH, "//a[@class='sidenav-icon-row' and text()='Change Benefits']")
+
+    def set_change_benefits(self):
+        self.get_change_benefits().click()
+        
+        Driver.Instance.switch_to.window(Driver.Instance.window_handles[-1])
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply pages/home_page.py]: Yes, when we select a framework they should have a better method for switching to the current window.

File: runner.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+
+import argparse
+import unittest2 as unittest
+from lib.helpers import get_modules
+import nose
+
+class Runner(object):
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply runner.py]: docker and docker-compose will be a component of framework but not the controller. Again, this is very basic for any framework. You want the ability to control what components you call and use not be controlled by a third party application. Using docker's interface directly is terse at best. You mentioned to me that some people commented that the docket cli was too complicated.

File: tests/test_benenroll_happypath.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+class BenEnrollHappyPath(namely_unittest.TestCase):
+    # Calls the default setUp
+    def setUp(self):
+        super().setUp()
+
+    # Tests
+    def test_benenroll_happypath(self):
+        login_page = LoginPage()
+        login_page.goto(self.test_data["company"])
```
[the remark this answers is not in the document]
Comment (IsiahG) [reply tests/test_benenroll_happypath.py]: Not sure how to answer that one...

File: lib/ben_enroll_base_page.py (hunk 1 above)
Comment (IsiahG) [reply lib/ben_enroll_base_page.py]: I never remember committing to composition... The design is solid.

File: pom_generate.py (hunk 2 above)
Comment (IsiahG) [reply pom_generate.py]: Thanks.

**Pull request merged**